Sources/PackageCollectionsSigning/PackageCollectionSigning.swift

```diff
@@ -8,43 +8,127 @@
  */
 
+import Dispatch
 import Foundation
 
+import Basics
 import PackageCollectionsModel
+import TSCBasic
 
-public struct PackageCollectionSigning {
+public protocol PackageCollectionSigner {
+    /// Signs package collection using the given certificate and key.
+    ///
+    /// - Parameters:
+    ///   - collection: The package collection to be signed
+    ///   - certChainPaths: Paths to all DER-encoded certificates in the chain. The certificate used for signing
+    ///                     must be the first in the array.
+    ///   - certPrivateKeyPath: Path to the private key (*.pem) of the certificate
+    ///   - certPolicyKey: The key of the `CertificatePolicy` to use for validating certificates
+    ///   - callback: The callback to invoke when the signed collection is available.
+    func sign(collection: PackageCollectionModel.V1.Collection,
+              certChainPaths: [URL],
+              certPrivateKeyPath: URL,
+              certPolicyKey: CertificatePolicyKey,
+              callback: @escaping (Result<PackageCollectionModel.V1.SignedCollection, Error>) -> Void)
+}
+
+public protocol PackageCollectionSignatureValidator {
+    /// Validates a signed package collection.
+    ///
+    /// - Parameters:
+    ///   - signedCollection: The signed package collection
+    ///   - certPolicyKey: The key of the `CertificatePolicy` to use for validating certificates
+    ///   - callback: The callback to invoke when the result is available.
+    func validate(signedCollection: PackageCollectionModel.V1.SignedCollection,
+                  certPolicyKey: CertificatePolicyKey,
+                  callback: @escaping (Result<Void, Error>) -> Void)
+}
+
+// MARK: - Implementation
+
+public struct PackageCollectionSigning: PackageCollectionSigner, PackageCollectionSignatureValidator {
     public typealias Model = PackageCollectionModel.V1
 
     private static let minimumRSAKeySizeInBits = 2048
 
-    let certPolicy: CertificatePolicy
+    /// URL of the optional directory containing root certificates to be trusted.
+    private let trustedRootCertsDir: URL?
+    /// Root certificates to be trusted in additional to those found in `trustedRootCertsDir`
+    private let additionalTrustedRootCerts: [Certificate]?
+
+    /// The `DispatchQueue` to use for callbacks
+    private let callbackQueue: DispatchQueue
+    /// Diagnostics engine to emit warnings and errors
+    private let diagnosticsEngine: DiagnosticsEngine
+
+    /// Internal cache/storage of `CertificatePolicy`s
+    private let certPolicies: [CertificatePolicyKey: CertificatePolicy]
 
-    public init() {
-        self.init(certPolicy: NoopCertificatePolicy())
+    private let encoder: JSONEncoder
+    private let decoder: JSONDecoder
+
+    public init(trustedRootCertsDir: URL? = nil, additionalTrustedRootCerts: [String]? = nil, callbackQueue: DispatchQueue = DispatchQueue.global(), diagnosticsEngine: DiagnosticsEngine = DiagnosticsEngine()) {
```

Comment:
similarly it would be better not to default diagnosticsEngine but rather have it set by the caller. the only place we might consider having a default is in the top level PackageCollections